ValueError: unsupported protocol with HTTP/1.0 proxy server

When using websockets 15.0.1 to establish a connection through a HTTP/1.0 only proxy server, I get the following traceback:
```
venv/lib64/python3.11/site-packages/websockets/asyncio/client.py:765: in run_parser
    next(self.parser)
venv/lib64/python3.11/site-packages/websockets/http11.py:250: in parse
    raise ValueError(
E   ValueError: unsupported protocol; expected HTTP/1.1: HTTP/1.0 200 Connection Established

The above exception was the direct cause of the following exception:
tests/test_notification.py:31: in test_websocket
    async with websockets.connect(uri, origin = gateway_url, additional_headers = additional_headers, ssl = ssl_context, proxy = proxy) as websocket:
venv/lib64/python3.11/site-packages/websockets/asyncio/client.py:587: in __aenter__
    return await self
venv/lib64/python3.11/site-packages/websockets/asyncio/client.py:541: in __await_impl__
    self.connection = await self.create_connection()
venv/lib64/python3.11/site-packages/websockets/asyncio/client.py:440: in create_connection
    transport = await connect_http_proxy(
venv/lib64/python3.11/site-packages/websockets/asyncio/client.py:815: in connect_http_proxy
    await protocol.response
E   websockets.exceptions.InvalidProxyMessage: did not receive a valid HTTP response from proxy
```

Manually checking the behavior of the proxy server:
```bash
$ nc proxy-host.com 3128
CONNECT websocket-host.com:443 HTTP/1.1
Host: websocket-host.com

HTTP/1.0 200 Connection Established
```

It is my understanding [reading the spec](https://httpwg.org/specs/rfc9112.html#http.version) that responding with HTTP/1.0 to a HTTP/1.1 request is allowed.

--- a/docs/project/changelog.rst
+++ b/docs/project/changelog.rst
@@ -32,6 +32,11 @@ notice.
 
 *In development*
 
+Improvements
+............
+
+* Added support for HTTP/1.0 proxies.
+
 15.0.1
 ------
 

--- a/src/websockets/asyncio/client.py
+++ b/src/websockets/asyncio/client.py
@@ -754,8 +754,7 @@ class HTTPProxyConnection(asyncio.Protocol):
             self.reader.read_line,
             self.reader.read_exact,
             self.reader.read_to_eof,
-            include_body=False,
-            allow_http10=True,
+            proxy=True,
         )
 
         loop = asyncio.get_running_loop()

--- a/src/websockets/asyncio/client.py
+++ b/src/websockets/asyncio/client.py
@@ -755,6 +755,7 @@ class HTTPProxyConnection(asyncio.Protocol):
             self.reader.read_exact,
             self.reader.read_to_eof,
             include_body=False,
+            allow_http10=True,
         )
 
         loop = asyncio.get_running_loop()

--- a/src/websockets/http11.py
+++ b/src/websockets/http11.py
@@ -213,8 +213,7 @@ class Response:
         read_line: Callable[[int], Generator[None, None, bytes]],
         read_exact: Callable[[int], Generator[None, None, bytes]],
         read_to_eof: Callable[[int], Generator[None, None, bytes]],
-        include_body: bool = True,
-        allow_http10: bool = False,
+        proxy: bool = False,
     ) -> Generator[None, None, Response]:
         """
         Parse a WebSocket handshake response.
@@ -250,7 +249,7 @@ class Response:
             protocol, raw_status_code, raw_reason = status_line.split(b" ", 2)
         except ValueError:  # not enough values to unpack (expected 3, got 1-2)
             raise ValueError(f"invalid HTTP status line: {d(status_line)}") from None
-        if allow_http10:  # some proxies still use HTTP/1.0
+        if proxy:  # some proxies still use HTTP/1.0
             if protocol not in [b"HTTP/1.1", b"HTTP/1.0"]:
                 raise ValueError(
                     f"unsupported protocol; expected HTTP/1.1 or HTTP/1.0: "
@@ -277,12 +276,12 @@ class Response:
 
         headers = yield from parse_headers(read_line)
 
-        if include_body:
+        if proxy:
+            body = b""
+        else:
             body = yield from read_body(
                 status_code, headers, read_line, read_exact, read_to_eof
             )
-        else:
-            body = b""
 
         return cls(status_code, reason, headers, body)
 

--- a/src/websockets/http11.py
+++ b/src/websockets/http11.py
@@ -214,6 +214,7 @@ class Response:
         read_exact: Callable[[int], Generator[None, None, bytes]],
         read_to_eof: Callable[[int], Generator[None, None, bytes]],
         include_body: bool = True,
+        allow_http10: bool = False,
     ) -> Generator[None, None, Response]:
         """
         Parse a WebSocket handshake response.
@@ -249,10 +250,17 @@ class Response:
             protocol, raw_status_code, raw_reason = status_line.split(b" ", 2)
         except ValueError:  # not enough values to unpack (expected 3, got 1-2)
             raise ValueError(f"invalid HTTP status line: {d(status_line)}") from None
-        if protocol != b"HTTP/1.1":
-            raise ValueError(
-                f"unsupported protocol; expected HTTP/1.1: {d(status_line)}"
-            )
+        if allow_http10:  # some proxies still use HTTP/1.0
+            if protocol not in [b"HTTP/1.1", b"HTTP/1.0"]:
+                raise ValueError(
+                    f"unsupported protocol; expected HTTP/1.1 or HTTP/1.0: "
+                    f"{d(status_line)}"
+                )
+        else:
+            if protocol != b"HTTP/1.1":
+                raise ValueError(
+                    f"unsupported protocol; expected HTTP/1.1: {d(status_line)}"
+                )
         try:
             status_code = int(raw_status_code)
         except ValueError:  # invalid literal for int() with base 10

--- a/src/websockets/sync/client.py
+++ b/src/websockets/sync/client.py
@@ -498,8 +498,7 @@ def read_connect_response(sock: socket.socket, deadline: Deadline) -> Response:
         reader.read_line,
         reader.read_exact,
         reader.read_to_eof,
-        include_body=False,
-        allow_http10=True,
+        proxy=True,
     )
     try:
         while True:

--- a/src/websockets/sync/client.py
+++ b/src/websockets/sync/client.py
@@ -499,6 +499,7 @@ def read_connect_response(sock: socket.socket, deadline: Deadline) -> Response:
         reader.read_exact,
         reader.read_to_eof,
         include_body=False,
+        allow_http10=True,
     )
     try:
         while True:

--- a/tests/test_http11.py
+++ b/tests/test_http11.py
@@ -333,6 +333,23 @@ class ResponseTests(GeneratorTestCase):
         response = self.assertGeneratorReturns(self.parse(include_body=False))
         self.assertEqual(response.body, b"")
 
+    def test_parse_http10(self):
+        self.reader.feed_data(b"HTTP/1.0 200 OK\r\nContent-Length: 0\r\n\r\n")
+        response = self.assertGeneratorReturns(self.parse(allow_http10=True))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.reason_phrase, "OK")
+        self.assertEqual(response.body, b"")
+
+    def test_parse_http10_unsupported_protocol(self):
+        self.reader.feed_data(b"HTTP/1.2 400 Bad Request\r\n\r\n")
+        with self.assertRaises(ValueError) as raised:
+            next(self.parse(allow_http10=True))
+        self.assertEqual(
+            str(raised.exception),
+            "unsupported protocol; expected HTTP/1.1 or HTTP/1.0: "
+            "HTTP/1.2 400 Bad Request",
+        )
+
     def test_serialize(self):
         # Example from the protocol overview in RFC 6455
         response = Response(

--- a/tests/test_http11.py
+++ b/tests/test_http11.py
@@ -328,22 +328,22 @@ class ResponseTests(GeneratorTestCase):
         response = self.assertGeneratorReturns(self.parse())
         self.assertEqual(response.body, b"")
 
-    def test_parse_without_body(self):
+    def test_parse_proxy_response_does_not_read_body(self):
         self.reader.feed_data(b"HTTP/1.1 200 Connection Established\r\n\r\n")
-        response = self.assertGeneratorReturns(self.parse(include_body=False))
+        response = self.assertGeneratorReturns(self.parse(proxy=True))
         self.assertEqual(response.body, b"")
 
-    def test_parse_http10(self):
+    def test_parse_proxy_http10(self):
         self.reader.feed_data(b"HTTP/1.0 200 OK\r\nContent-Length: 0\r\n\r\n")
-        response = self.assertGeneratorReturns(self.parse(allow_http10=True))
+        response = self.assertGeneratorReturns(self.parse(proxy=True))
         self.assertEqual(response.status_code, 200)
         self.assertEqual(response.reason_phrase, "OK")
         self.assertEqual(response.body, b"")
 
-    def test_parse_http10_unsupported_protocol(self):
+    def test_parse_proxy_unsupported_protocol(self):
         self.reader.feed_data(b"HTTP/1.2 400 Bad Request\r\n\r\n")
         with self.assertRaises(ValueError) as raised:
-            next(self.parse(allow_http10=True))
+            next(self.parse(proxy=True))
         self.assertEqual(
             str(raised.exception),
             "unsupported protocol; expected HTTP/1.1 or HTTP/1.0: "

